pageserver/src/tenant/blob_io.rs

```diff
@@ -117,14 +123,28 @@ impl<const BUFFERED: bool> BlobWriter<BUFFERED> {
 
     const CAPACITY: usize = if BUFFERED { PAGE_SZ } else { 0 };
 
-    #[inline(always)]
     /// Writes the given buffer directly to the underlying `VirtualFile`.
     /// You need to make sure that the internal buffer is empty, otherwise
     /// data will be written in wrong order.
-    async fn write_all_unbuffered(&mut self, src_buf: &[u8]) -> Result<(), Error> {
-        self.inner.write_all(src_buf).await?;
-        self.offset += src_buf.len() as u64;
-        Ok(())
+    #[inline(always)]
+    async fn write_all_unbuffered<B: BoundedBuf>(
+        &mut self,
+        src_buf: B,
+    ) -> (B::Buf, Result<(), Error>) {
+        let src_buf_len = src_buf.bytes_init();
+        let (src_buf, res) = if src_buf_len > 0 {
```

Comment:
I don't understand why we need this if here: are empty `Slice`s forbidden by tokio-epoll-uring? Reading [the code](https://github.com/neondatabase/tokio-epoll-uring/blob/main/uring-common/src/buf/slice.rs) this doesn't seem to be the case. I think we can just use the "then" case always.

Reply:
See the PR description for why it's needed:

> NB: Invoking BoundedBuf::slice(..) will return a slice that includes the uninitialized portion of BoundedBuf.
> I.e., the portion between bytes_init() and bytes_total().
> It's a safe API that actually permits access to uninitialized memory.
> Not great.